internal/cluster/watch.go

```diff
@@ -89,50 +99,31 @@ func containsFactForURL(pipelineActivity *jenkinsv1.PipelineActivity, url string
 	return false
 }
 
-func updatePipelineActivity(name string, namespace string, fact jenkinsv1.Fact, jxClient *jenkinsclientv1.JenkinsV1Client) {
-	var pipelineActivity *jenkinsv1.PipelineActivity
-	var err error
-	// re-get the pipeline activity
+func updatePipelineActivity(name string, namespace string, fact jenkinsv1.Fact, pipelineActivitiesGetter jenkinsclientv1.PipelineActivitiesGetter) error {
 	f := func() error {
-		pipelineActivity, err = jxClient.PipelineActivities(namespace).Get(name, metav1.GetOptions{})
+		pipelineActivity, err := pipelineActivitiesGetter.PipelineActivities(namespace).Get(name, metav1.GetOptions{})
 		if err != nil {
 			return err
 		}
-		return nil
-	}
-
-	err = util.ApplyWithBackoff(f)
-	if err != nil {
-		logger.Errorf("error retrieving PipelineActivity %s with uuid %s for update: %s", pipelineActivity.Name, pipelineActivity.UID, err)
-		return
-	}
 
-	found, index, err := indexOfJacocoCoverageFactType(pipelineActivity)
-	if err != nil {
-		logger.Error(err)
-	}
+		found, index, err := indexOfJacocoCoverageFactType(pipelineActivity)
+		if err != nil {
+			return backoff.Permanent(err)
+		}
 
-	if found {
-		// Updating the existing jacoco report with the new one. Is this what we want? (HF)
-		pipelineActivity.Spec.Facts[index] = fact
-	} else {
-		pipelineActivity.Spec.Facts = append(pipelineActivity.Spec.Facts, fact)
-	}
+		if found {
+			pipelineActivity.Spec.Facts[index] = fact
+		} else {
+			pipelineActivity.Spec.Facts = append(pipelineActivity.Spec.Facts, fact)
+		}
 
-	// update the pipeline CRD
-	f = func() error {
-		_, err = jxClient.PipelineActivities(pipelineActivity.Namespace).Update(pipelineActivity)
+		_, err = pipelineActivitiesGetter.PipelineActivities(pipelineActivity.Namespace).Update(pipelineActivity)
 		if err != nil {
 			return err
 		}
 		return nil
 	}
-	err = util.ApplyWithBackoff(f)
-	if err != nil {
-		logger.Errorf("error updating PipelineActivity %s: %s", pipelineActivity.Name, err)
-	} else {
-		logger.Infof("successfully updated PipelineActivity %s with data from %s", pipelineActivity.Name, fact.Original.URL)
-	}
+	return util.ApplyWithBackoff(f)
```

Comment:
I guess the question is just whether `ApplyWithBackoff` is actually written correctly and able to handle this use case. IMO it should

* succeed on 2xx status
* maybe handle 3xx series somehow (not sure if this is applicable in K8s APIs)
* retry with exponential backoff upon 5xx errors
* retry immedately on 409
* fail on any other 4xx